src/compiler/checker.ts

```diff
@@ -14845,6 +14841,17 @@ namespace ts {
                 }
             }
 
+            // Referencing Abstract Properties within Constructors is not allowed
+            if ((flags & ModifierFlags.Abstract) && symbolHasNonMethodDeclaration(prop)) {
+                const declaringClassDeclaration = <ClassLikeDeclaration>getClassLikeDeclarationOfSymbol(getParentOfSymbol(prop));
+                const declaringClassConstructor = declaringClassDeclaration && findConstructorDeclaration(declaringClassDeclaration);

```

Comment:
I think it would be simpler to just walk up from the property access node looking for a constructor:

```ts
if (declaringClassDeclaration && 
    findAncestor(node, n => n === declaringClassDeclaration ? "quit" : isConstructorDeclaration(n) && nodeIsPresent(n.body)) {
  error(...
```

The code as it stands has a linear scan of method declarations, followed by a log-linear walk of ancestors that ends at the source file. The code above just has a log-linear walk of ancestors that ends at the class declaration.

Reply:
The only concern I have with this is that it might throw too many false-positives. The existing `isNodeWithinFunction` only compares one level up, it doesn't do a walk of ancestors because that would incorrectly throw errors in code such as:

```
abstract class A {
    constructor() {
        this.method = () => {
            return this.property;
        }
    }
    abstract property: string;
}
```

This code is fine since the abstract property isn't actually being accessed in the constructor, but if we do a walk of the ancestors then it would see `this.property` access as being inside the constructor and emit an error.

Perhaps a solution would be to also have the `findAncestor` short-circuit on a non-constructor `FunctionLike` node? So that accessing an abstract property within a function that's in the constructor is okay while accessing it in some other way isn't.

Reply:
Ah, yes, I didn't think about that. Short-circuiting the walk when you hit a function-like node is a good idea.